pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,32 +16,134 @@
 
 // If you feel like getting in touch with us, you can do so at info@botlabs.org
 
-use crate::{linkable_account::LinkableAccountId, AccountIdOf, Config, ConnectionRecordOf, DidIdentifierOf, Pallet};
-
-use crate::{ConnectedAccounts as ConnectedAccountsV2, ConnectedDids as ConnectedDidsV2};
+use crate::{
+	linkable_account::LinkableAccountId, migration_state::MigrationState, AccountIdOf, Config,
+	ConnectedAccounts as ConnectedAccountsV2, ConnectedDids as ConnectedDidsV2, ConnectionRecordOf, DidIdentifierOf,
+	MigrationStateStore, Pallet,
+};
 
+use codec::{Decode, Encode, MaxEncodedLen};
 use frame_support::{
+	storage::KeyPrefixIterator,
 	storage_alias,
 	traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
-	Blake2_128Concat,
-};
-use sp_std::{marker::PhantomData, vec};
-
-#[cfg(feature = "try-runtime")]
-use {
-	frame_support::{
-		codec::{Decode, Encode},
-		inherent::Vec,
-	},
-	sp_runtime::SaturatedConversion,
+	Blake2_128Concat, ReversibleStorageHasher, StoragePrefixedMap,
 };
+use scale_info::TypeInfo;
+use sp_runtime::{AccountId32, DispatchError};
+use sp_std::marker::PhantomData;
 
+use sp_std::vec::Vec;
+
+/// Keytype changed from `AccountId` to `LinkableAccountId` changed in V3
 #[storage_alias]
 type ConnectedDids<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, AccountIdOf<T>, ConnectionRecordOf<T>>;
+/// Second keytype changed from `AccountId` to `LinkableAccountId` changed in V3
 #[storage_alias]
 type ConnectedAccounts<T: Config> =
 	StorageDoubleMap<Pallet<T>, Blake2_128Concat, DidIdentifierOf<T>, Blake2_128Concat, AccountIdOf<T>, ()>;
 
+#[derive(Clone, Debug, Eq, PartialEq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum MixedStorageKey {
+	V1(AccountId32),
+	V2(LinkableAccountId),
+}
+
+/// Get an iterator that distinguishes between new and old storage keys.
+///
+/// This is only possible because old storage keys are 32 bytes long and new
+/// storage keys either 21 or 33 bytes.
+pub(crate) fn get_mixed_storage_iterator<T: Config>(
+	previous_key: Option<Vec<u8>>,
+) -> KeyPrefixIterator<MixedStorageKey> {
+	let previous_key = previous_key.unwrap_or_else(|| ConnectedDids::<T>::final_prefix().to_vec());
+	KeyPrefixIterator::new(
+		ConnectedDids::<T>::final_prefix().to_vec(),
+		previous_key,
+		|raw_key_without_prefix| {
+			let mut key_material = Blake2_128Concat::reverse(raw_key_without_prefix);
```

Comment:
I'm not 100% sure on this one to be honest, since the keys are already of the correct length. But the prefix is only the Pallet and Map name. `raw_key_without_prefix` is the key inside the storage map and this one is still `Blake2_128Concat` "hashed" what ever that means.

Reply:
What do you mean the prefix is "only" the pallet and map name? What else could there be in there? If `ConnectedDids` would be a storage nmap, it would make sense to split the first key from the rest, basically. In this case it's a map, so I don't see what else could be part of `raw_key_without_prefix` beyond the actual `LinkableAccountId` or `AccountId` in the old version. Just asking out of curiosity to understand better what's the need for this operation here.

Reply:
But the `LinkableAccountId`/`AccountId` are still hashed. If you look at the storage declaration: `StorageMap<Pallet<T>, Blake2_128Concat, AccountIdOf<T>, ConnectionRecordOf<T>>` there is this `Blake2_128Concat`. This basically hashes the `LinkableAccountId`/`AccountId`. So we need to revert the hashing.

Reply:
I see. I thought the key was simply hashed, and not also appended itself. Clear now!